docs/rfcs/0007-flow-modules.md

```diff
@@ -0,0 +1,295 @@
+
+### Add additional capabilities to load strings
+
+Modules will not care about the source of the string that represents the river syntax, nor will modules have any inherent reload semantics. The component supplying the string will be responsible for the source and will notify the module when the input changes so that it can utilize the new river configuration.
+
+### Any type of versioning
+
+Modules will not contain any sort of versioning nor will check for compatibility outside the normal river checks.
+
+### Any user interface work beyond ensuring it works as the UI currently does
+
+Users will not be able to drill into modules, they will be represented as any other normal component. 
+
+## Example
+
+```river
+// module
+argument "password" {
+    optional = false
+    comment = "password for mysql"
+}
+
+argument "username" {
+    optional = false
+    comment = "username for mysql"
+}
+
+export "targets" {
+    comment = "targets for the integration"
+    value = integrations.mysql.server1.targets
+}
+
+integrations.mysql "server1" {
+    username = argument.username.value
+    password = argument.password.value
+}
+
+```
+
+```river
+// parent
+
+local.file "mysql" {
+    filename = "/test/mysql.river"
+}
+
+module.single "mysql" {
+    content = local.file.mysql.content
+    arguments = {
+        {
+            "password" = PASSWORD,
+            "username" = USERNAME,
+        }
+    }
+}
+
+prometheus.scrape "scraper" {
+    targets = module.single.mysql.exports.targets
+}
+
+```
+
+## Limitations
+
+* Duplicate modules cannot be nested, this may or may not be enforced by the system
+* Singleton components are not supported at this time. Example [node_exporter](https://grafana.com/docs/agent/latest/flow/reference/components/prometheus.integration.node_exporter/).
+* Modules will not prevent competing resources, such as starting a server on the same port
+* [Configuration blocks](https://grafana.com/docs/agent/latest/flow/reference/config-blocks/#configuration-blocks) will not be supported. 
+* Arguments and exports within a module must be unique
+
+## Proposal
+
+Add the ability to load `modules` as subgraphs to the primary `graph`. Modules may call other modules within a reasonable stack size depth. Modules are represented as a river string that is interpreted with a defined set of arguments and exports.
+
+The initial component will be `module.single` that will load a single module. Internally these modules will be namespaced so they cannot affect children or parent graphs except via arguments and exports.
+
+Modules will have access to any standard function and any other component exempting singletons. Internally each component in the module will have an `id` that is prepended with the parent's `id` for identification purposes outside of the module. Within the module a component can reference another sibling component normally. There are no known limits on the datatype that a module can use as an argument or export.
```

Comment:
It might help to put an example of the `id`  you mention here. Correct me if I'm wrong, but how this reads to me is if I had the following graph of module dependencies:
`A -> B -> C -> D`
Then a component in module `D`, let's call it a `discovery.relabel "relabeler"` would have an `id` field on it that looks like `ABCD`. 

The word 'internally' is throwing me of here. Is this `id` intended to be used by the end user to reference/identify a component nestled deep in a module dependency path, or is this just for debugging/logging?

Reply:
The configuration you write does not care but the only reason I included it is because of metric naming and passing the data path for components to use. So for a metric you may have `failed_requests{id="module.single.loki.loki.receiver.syslog.mysyslogreceiver"} 0`. In this case the `module.single.loki "loki"` module loads component `loki.receiver.syslog "mysyslogreceiver"` The data path for any file system usage would be equally odd.